### DISCUSSION TITLE
SetPowerOutput(20, false) not working

### DISCUSSION THREAD

**Discussion by jacasais** [jgromes/RadioLib#983] [Q&A]
Dear all,

Good day

I am using radiolib with an SX1278 Module, i can transmit and receive without problem but we tested on our spectrum analizer the power output for the module and is alwas 10dBm, even that on the arduino code is set to: if (radio.setOutputPower(20, false) == RADIOLIB_ERR_INVALID_OUTPUT_POWER) {
    Serial.println(F("Selected output power is invalid for this module!"));
    while (true);
  }
Of course we do not get any error that the output power is not correct, unless we set it to for example 22, but even with this setting, is outputting only 10 dBm.
We use this pins on our SX1278:
SX1278 radio = new Module(7, 2, 9, 3);
Does anyone know how we can fix this?

Thank you.

**Comment by jgromes** [comment]
Since oyu have not posted your complete code (and ignored the issue template in the process), it's very hard to guess, but I will attempt to anyway: did you increase the overcurent protection limit? I'd guess not, it's set to 60 mA by default, not enough for +20 dBm output.

**Comment by jacasais** [comment]
Dear jgromes,
Sorry for deleting the template is the first time i post on GitHub and i did not really know how to proceed.
Thank you for your answer.
I have tried with this setting for the current limit: 
 if (radio.setOutputPower(20, false) == RADIOLIB_ERR_INVALID_OUTPUT_POWER) {
    Serial.println(F("Selected output power is invalid for this module!"));
    while (true);
  }
  if (radio.setCurrentLimit(240) == RADIOLIB_ERR_INVALID_CURRENT_LIMIT) {
    Serial.println(F("Selected current limit is invalid for this module!"));
    while (true);
  }
But still only getting 10 dBm, the module is XL1278-SMT maybe is somethig to do with this?

Thank you for your answer highly appreciated

**Comment by jgromes** [comment]
I am still missing basic information, such as the code, which Arduino you are using etc.

> the module is XL1278-SMT

No idea what that is. Quick google search several different Chinese modules. Do you know whether it uses the RFO or the PA_BOOST pin? Is the power source sufficient? How are you measuring the output?

**Comment by jacasais** [comment]
```c++
#include <RadioLib.h>
#include <WiFi.h>
#include <WebServer.h>
#include <SPIFFS.h>

// Configuración WiFi y servidor web
const char* ssid = "wifi";
const char* password = "123456789";
WebServer server(80);
//unsigned long wifiStartTime;
//const unsigned long wifiTimeout = 60000; // 60 segundos
bool isWifiActive = false;
unsigned long lastRequestTime = 0;
const unsigned long wifiTimeout = 60000; // 60 segundos
String lastConfigData = ""; //variable para leer los datos desde la pagina admin pulsando Leer

#define LED_PIN 1 
#define SENSOR_PIN 10 
unsigned long previousMillis = 0; // Variable para almacenar la última vez que se actualizó el LED
const long intervaloEncendido = 300; // Medio segundo
const long intervaloApagado = 2000; // 2 segundos

// Configuración del módulo LoRa SX1278
SX1278 radio = new Module(7, 2, 9, 3);
bool mensajeLoRaEnviado = false;



float frecuencia = 433.0; // Frecuencia inicial
float anchoBanda = 62.5; // Ancho de banda inicial
String SerialNumber = "433T0000101"; //Numero de serie ejemplo
String ID = "01"; //ID ejemplo maximo ID 99 


void setup() {
 if (!SPIFFS.begin(true)) {
        Serial.println("Error al montar el sistema de archivos");
        return;
    }

    server.on("/", HTTP_GET, handleRoot);

    // Agregar un punto de acceso para servir el archivo de imagen
    server.on("/pickpoint.png", HTTP_GET, []() {
        File file = SPIFFS.open("/pickpoint.png", "r");
        if (!file) {
            server.send(404, "text/plain", "Archivo no encontrado");
            return;
        }
        server.streamFile(file, "image/png");
        file.close();
    });  
// Configurar el servidor web para procesar datos y ademas
// Configura el ESP32 como punto de acceso
  WiFi.softAP(ssid, password);
  isWifiActive = true;
  IPAddress IP = WiFi.softAPIP();
  Serial.print("IP del AP: ");
  Serial.println(IP);
  server.on("/", HTTP_GET, handleRoot);
  server.on("/admin", HTTP_GET, handleAdminLoginForm);
  server.on("/admin", HTTP_POST, handleAdmin);
  server.on("/program", HTTP_POST, handleProgram);
  server.on("/setFrequency", HTTP_GET, []() {
    String frequency = server.arg("frequency");
    procesarComando("F" + frequency);
    server.send(200, "text/plain", "Frecuencia programada: " + frequency);
  });

  server.begin();
  pinMode(A0, INPUT); // Configurar el pin A0 para la lectura de voltaje
  pinMode(LED_PIN, OUTPUT); // Configura el pin del LED como salida
  pinMode(SENSOR_PIN, INPUT); // Configura el pin del sensor como entrada
  digitalWrite(LED_PIN, LOW); // Asegúrate de que el LED esté apagado al inicio

  Serial.begin(9600);

  Serial.print(F("[SX1278] Initializing ... "));
  int state = radio.begin(frecuencia, anchoBanda);
    // Configurar el Factor de Dispersión (Spreading Factor)
   // state = radio.setSpreadingFactor(12);
   // if (state != RADIOLIB_ERR_NONE) {
   //   Serial.println(F("Falló la configuración de SF!"));
   // }
    // Configurar el Código de Tasa (Coding Rate)
   // state = radio.setCodingRate(7);
   // if (state != RADIOLIB_ERR_NONE) {
   //   Serial.println(F("Falló la configuración de CR!"));
   // }
  if (radio.setOutputPower(20, false) == RADIOLIB_ERR_INVALID_OUTPUT_POWER) {
    Serial.println(F("Selected output power is invalid for this module!"));
    while (true);
  }
  if (radio.setCurrentLimit(240) == RADIOLIB_ERR_INVALID_CURRENT_LIMIT) {
    Serial.println(F("Selected current limit is invalid for this module!"));
    while (true);
  }
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("Initialization successful!"));
    digitalWrite(GPS, HIGH);
  } else {
    Serial.print(F("Initialization failed, code "));
    Serial.println(state);
    while (true);
  }

}
```

I use an ESP32C3Mini module connected to the XL1278-SMT lora module, i can transmit, no problem, the only problem is that i connect the lora to my spectrum analyzer to check the power output and is alwas 10 db no matter what i set on the code. The current is not the problem the lora gets the power from a 18650 battery so this is not the problem. On the specifications for this module, says that can transmitt up to 20,5 dBm.

Thank you sir.

**Comment by jgromes** [comment]
> spectrum analyzer to check the power output and is alwas 10 db

10 dB is just a relative measure, what is the reference level? And what is the measurement setup? Spectrum analyzers are typically quite sensitive to input power and will get damaged if you exceed its limit, do you have some attenuators in path?

> The current is not the problem the lora gets the power from a 18650 battery

So you have the LoRa module connected directly to the battery? Seems unlikely, but if it really is the case it will most definitely destroy it, the absolute maximum the module can tolerate is 3.9V, while a fully charged Li-ion cell will be at 4.2V.

I have measured power output from SX1278 using RF power meter - not spectrum analyzer, an actual RF power meter with proper atteniuation. The output was as expected, so I am fairly confident it is not an issue in the library, but rather with your setup, or measurement, so I will convert this to a discussion.

**Comment by jacasais** [comment]
Dear jromes

The battery is set to 3,6 Volts, i know fully charged is 4.2 Volts and that is not good.
About the spectrum analizer, our Marconi allow 4 W direct input so we do not need to add any attenuator for this type of power, and we get exactly 10dB output, no matter is we set power to 20, 10, 18,... does not make any difference, could you recommend me an sx1278 module that you already tested so i can buy it and compare?

Thank you.
Best Regards

**Reply by jgromes** [reply]
> could you recommend me an sx1278 module that you already tested so i can buy it and compare?

I used this one, in the 433 MHz version: https://www.lilygo.cc/products/t-beam-v1-1-esp32-lora-module

**Comment by jacasais** [comment]
![POWER SET](https://github.com/jgromes/RadioLib/assets/64363833/e188b4d9-6907-4ca5-b963-f89ff0e6e8de)
This is what i read on the SX1278 Manual, they talk about RegPaDac address 0x4d, but well, my point is more electronics, i do not know about firmware so much or libraries, sorrey.
Hope this could be usefull.

Thanks

**Reply by jgromes** [reply]
> This is what i read on the SX1278 Manual, they talk about RegPaDac address 0x4d

This is already being handled by RadioLib:

https://github.com/jgromes/RadioLib/blob/0ea00fad44ffdadd3279d8fdd66fc259df8879f0/src/modules/SX127x/SX1278.cpp#L333-L337

**Reply by jacasais** [reply]
Ok, so I realise that inside the library folders there is that SX1278.h file, but do I need to do something on the Arduino code to force it to use that file? My doubt is maybe if is using another file for example SX127x.h or this is automatically detected by the library?

Thank you

**Reply by jgromes** [reply]
> do I need to do something on the Arduino code to force it to use that file?

No.

> this is automatically detected by the library?

In your code, you have created `radio` as an instance of the `SX1278` class. Had the library not included SX1278.h and compiled SX1278.cpp, that class would not be defined, leaving you with a compilation error. Since you saw no such error, it means that both files are managed correctly.

How did you validate your meassurement setup? Did you measure agains a known good power source? Or at the very least, included a known attenuation in the RF path? If you add e.g. a 3 dB attenuator and your equipment shows a 3 dB drop, then at least the relative measurements are correct.

**Reply by jacasais** [reply]
Please find attached the picture of the spectrum analizer.

This value is always the same nevermind if I set on the code 20, or 10, or 18....

Thanks 
![image](https://github.com/jgromes/RadioLib/assets/64363833/af3a1853-92bd-45b8-989a-405e06438ac7)

**Reply by jgromes** [reply]
@jacasais just to satisfy my own curiosity, I got my power meter and verified that I can indeed get over +10 dBm output from SX1278. Granted, my measurements are not very precise, but I can clearly see the output power rising in roughly 1 dBm steps all the way in the supported range for SX1278, from -2 to +17 and then jumping to +20 dBm. So I am very confident that this is not an issue in the library, but rather with your setup. I don't know if it's your measurement or the LoRa module itself, but I would start by verification of the measurement setup with a known reference.

**Reply by jacasais** [reply]
@jgromes good day.

Could you let me know the brand and model of your power meter?

We tested with three different spectrum analyzers and we set it at spam 100khz this is to make a more accurate reading because we use 62.5 kHz band width, and the result is the picture I uploaded previously, 10db, with the outputPower set to 20.
If we set a bigger spam on the analizer for example 1000khz we can read about 19 dB but the way to read it on this case is not accurate.

Our spectrum analizer is calibrated every year on a laboratory but apart from this we have a Marconi radio test set where we can generate a know output power of for example 20 db and we can read that on the spectrum analizer with the correct value, so we know the reading on this device is correct.
Any way if you tell me your power meter model I would like to buy the same one so I can compare.

I have one more question, on the program I did not set SF or CR so I wonder by default if I do not set a value, what is the SF and CR that will be used if not declared?

Thank you for your time.
Best Regards

**Reply by jgromes** [reply]
> Could you let me know the brand and model of your power meter?

A cheapo, no-brand chinese RF power meter, looks like this one: https://www.ebay.com/itm/374988291785 It's not very accurate, but the absolute value it reports doesn't really matter for my purposes. The important thing is that it reports increases of roughly 1 dBm each time I change the output power value in RadioLib. Because the bandwith in which it measures is completely unknown to me, I set the SX1278 to just send out unmodulated carrier with `transmitDirect`.

> Our spectrum analizer is calibrated every year on a laboratory but apart from this we have a Marconi radio test set where we can generate a know output power of for example 20 db and we can read that on the spectrum analizer with the correct value, so we know the reading on this device is correct.

I'm assuming you meant "20 dBm", otherwise the number is meaningless. Anyway, now since your measurement chain seems validated, it's time to investigate the other part - the SX127x. Specifically, how much power is it drawing? Are there any drops in its supply voltage?

> I have one more question, on the program I did not set SF or CR so I wonder by default if I do not set a value, what is the SF and CR that will be used if not declared?

See https://github.com/jgromes/RadioLib/wiki/Default-configuration

**Comment by HeadBoffin** [comment]
That design is rather old and may well have some of the foibles that the RFM95 had with actually co-operating with the attempts to change the power output under all circumstances. So you may want to look at using a far more recent design before disappearing down the rabbit hole of wondering why RadioLib isn't getting the results expected from the module - it can only send the commands, it's not unusual for some implementations to not do as they are asked.

As an aside, for a commercial product in a harsh environment, that module would be very very low down my list of choices on quality & power consumption for battery driven devices. Something based on SX1262 would be far more frugal. In the EU 20dBm would be illegal unless you have them floating more than 12 nautical miles from shore or in the Americas.

**Reply by serialrf433** [reply]
>  In the EU 20dBm would be illegal unless you have them floating more than 12 nautical miles from shore or in the Americas.

When you do not have a RF-licence and use the ISM regulations that can be used by anyone.
If you have for example in the EU a amateur license for transmitting RF, you can transmit legal with around over 55dBm on the same frequency.
There are many other license types that also could be used for many other different frequencies supported by the SX1278. Just writing down that 20dBm with an SX1278 is not legal in EU is not correct.

**Reply by HeadBoffin** [reply]
But it is in the ISM band which is what's being used here. If we had to cover every edge case we'd be lawyers.

**Reply by serialrf433** [reply]
You know the ISM band (433) and the amateur radio band (430 - 440) overlap completely? Its not some edge case. Its just the normal usage of this frequency around the world. You do not have to be 'something' to known things. You can be just a human (or Ai) and just know things.
https://en.wikipedia.org/wiki/ISM_radio_band
https://en.wikipedia.org/wiki/Amateur_radio_frequency_allocations
https://en.wikipedia.org/wiki/70-centimeter_band
